languages/elixir/exercises/concept/streams/.meta/example.ex

```diff
@@ -0,0 +1,20 @@
+defmodule Monopoly do
+  def d6 do
+    1..6
+    |> Enum.shuffle()
+    |> Enum.at(0)
+  end
+
+  def roll() do
+    fn -> {d6(), d6()} end
+    |> Stream.repeatedly()
+    |> Stream.with_index()
+    |> Stream.transform(false, fn {{d1, d2}, index}, acc ->
+      if index == 0 || (acc && index < 3) do
+        {[{d1, d2}], d1 == d2}
+      else
+        {:halt, false}
+      end
+    end)
+  end
```

Comment:
Yeah, I had my worries about the complexity of this, too. I thought I would find a `reduce` function for streams, thus building on top what we taught about `reduce` in the `enum` exercise, but there wasn't any. Maybe I could change the story somehow to avoid `Stream.transform`?

Reply:
My only idea at a simplification while still keeping the d6 roll story is to switch games from Monopoly to Mensch ärgere Dich nicht. There, you roll a single d6 and keep rolling as long as you roll 6s:

```
defmodule MenschAergereDichNicht do
  def d6 do
    1..6
    |> Enum.shuffle()
    |> Enum.at(0)
  end

  def roll(d6 \\ &d6/0) do
    d6
    |> Stream.repeatedly()
    |> Stream.transform(true, fn roll, acc ->
      if acc do
        {[roll], roll == 6}
      else
        {:halt, false}
      end
    end)
  end
end
```

But that doesn't remove `Stream.transform`.

Reply:
This example of transform is quite nice and I think it's quite a bit simpler. I didn't think the complexity came from the use of Stream.transform (it is the general case for many of the more niche transforms in the library like `reduce` for `Enum`). 

I think it was a big leap from just rolling a single die to rolling multiple dice up to multiple conditions.

`Stream.unfold` could be used too, but I think transform (like in your example) is more straightforward in this case:

```elixir
def roll(dice_fn) do                      
  Stream.unfold(dice_fn.(), fn
    :end -> nil
    6 -> {6, dice_fn.()}
    n -> {n, :end}
  end) |> Enum.to_list()
end
```

Reply:
> My only idea at a simplification while still keeping the d6 roll story is to switch games from Monopoly to Mensch ärgere Dich nicht.

Just to chip in. Another theme that we've used before is Dungeons & Dragons (of Wizards & Wizards) as we called, which also requires lots of use of dice (of various numbers of sides).